Colorspace plug override for ImageReader and ImageWriter

lucienfostier opened the pull request:
This pull request goal is to allow user to override the default colorspace picked by OpenImageIO to allow better control of the color workflow inside GafferImage.

( this is the first part, with just the ImageReader to start with )

- ImageReader:
  - Added "colorSpace" plug to allow user to override OIIO default colorspace.

- ImageReaderUI:
  - Added "colorSpace" plug preset.

lucienfostier commented:
Hey John,

this is the first part of the work on the colorspace plug for both ImageReader and ImageWriter.
I wanted to get your feedback before going any further that this was going into the right direction.

Cheers

johnhaddon commented on src/GafferImage/ImageReader.cpp:
```diff
@@ -83,10 +83,11 @@ ImageReader::ImageReader( const std::string &name )
 	addChild( new CompoundObjectPlug( "__intermediateMetadata", Plug::In, new CompoundObject, Plug::Default & ~Plug::Serialisable ) );
 	addChild( new StringPlug( "__intermediateColorSpace", Plug::Out, "", Plug::Default & ~Plug::Serialisable ) );
 	addChild( new ImagePlug( "__intermediateImage", Plug::In, Plug::Default & ~Plug::Serialisable ) );
-
```
Unnecessary formatting change.

johnhaddon commented on src/GafferImage/ImageReader.cpp:
```diff
@@ -292,15 +304,24 @@ void ImageReader::compute( ValuePlug *output, const Context *context ) const
 	if( output == intermediateColorSpacePlug() )
 	{
-		std::string intermediateSpace = "";
+		std::string userColorSpace = colorSpacePlug()->getValue();
```
I think the logic here can be simplified (and it needs to be changed somehow because all the tests are failing at present). My suggestion would be that if `colorSpacePlug->getValue() != ""`, we just use that as the result immediately, without even checking that it is a valid colorspace. If the user has entered an invalid space we need to alert them with an error (and the ColorSpace node should do this for us) rather than silently change behaviour.

johnhaddon commented on python/GafferImageUI/ImageReaderUI.py:
```diff
@@ -191,6 +199,21 @@
+
+			"description",
+			"""
+			The colour space of the input image.
```
Could we clarify here that this will be used to convert to linear space during loading?

johnhaddon commented on python/GafferImageUI/ImageReaderUI.py:
```diff
@@ -40,6 +40,14 @@
+
+def __colorSpacePresetValues( plug ) :
+
+	return IECore.StringVectorData( [ "oiioDefault" ] + GafferImage.OpenColorIOTransform.availableColorSpaces() )
```
This needs to be "" rather than "oiioDefault", so that it matches the actual default value of the plug, and the UI doesn't show the "Invalid" label.

lucienfostier commented:
Hey John,

I've pushed a few commits to address your comment. Let me know if you additional comments.

Cheers

johnhaddon commented on src/GafferImage/ImageReader.cpp:
```diff
@@ -103,7 +104,7 @@ ImageReader::ImageReader( const std::string &name )
 	colorSpace->inPlug()->setInput( oiioReader->outPlug() );
 	colorSpace->inputSpacePlug()->setInput( intermediateColorSpacePlug() );
-	colorSpace->outputSpacePlug()->setValue( OpenColorIO::ROLE_SCENE_LINEAR );
+	colorSpace->outputSpacePlug()->setValue( config->getColorSpace( OpenColorIO::ROLE_SCENE_LINEAR )->getName() );
```
Could you explain this one to me Lucien? What confuses me is that we seem to be using ROLE_SCENE_LINEAR directly here, and presumably getting away with it :

https://github.com/GafferHQ/gaffer/blob/master/startup/gui/ocio.py#L72

Should we be changing that line too, or is there a reason it's not needed there? I don't doubt that you've fixed something here, but I'd like to understand the situation better...

johnhaddon commented on src/GafferImage/ImageReader.cpp:
```diff
@@ -63,6 +63,7 @@ ImageReader::ImageReader( const std::string &name )
 	:	ImageNode( name )
 {
 	storeIndexOfNextChild( g_firstChildIndex );
+	OpenColorIO::ConstConfigRcPtr config = OpenColorIO::GetCurrentConfig();
```
Any reason to declare this so far away from the point it's actually used? I think putting it directly in front of the associated setValue() call would be clearer...

johnhaddon commented on src/GafferImage/ImageReader.cpp:
```diff
@@ -305,25 +306,27 @@ void ImageReader::compute( ValuePlug *output, const Context *context ) const
 	{
 		std::string userColorSpace = colorSpacePlug()->getValue();
-		std::string intermediateSpace = "oiioDefault";
+		std::string metadataSpace = "";
```
I might not have explained very well, but what I was trying to say was that if a userColorSpace has been specified, we don't even need to get the metadata and do the search through it - we can skip all that work entirely.

johnhaddon commented:
> I've pushed a few commits to address your comment. Let me know if you additional comments.

Thanks Lucien,
I've made a couple of comments and asked a question so I understand things better. If you want to address those and squash the fixes that'd be great, but they're not critical so if you don't get a chance by tomorrow, I'll take it as it is...
Thanks!
John

lucienfostier commented:
Hey John,

I've addressed your comment and squashed the commits.

Here is a screenshot attached to better explain the fix I made for the SCENE_LINEAR.
I don't know if you would consider that a bug because the computation is not the problem.
The problem is that the string plug doesnt contains the preset for the string value "scene_linear" but for OCIO this is a valid colorspace "identifier", thus the node was working fine in terms of output data.
I hope that clarifies things but let me know if not.

Cheers
![imagereadercolorspaceproblem](https://cloud.githubusercontent.com/assets/770272/24029260/1507174c-0a92-11e7-9ced-30c743a90a14.png)

johnhaddon commented:
Great - thanks Lucien!
